win32ss/user/winsrv/consrv/condrv/text.c

```diff
@@ -932,138 +933,224 @@ ConDrvReadConsoleOutputString(IN PCONSOLE Console,
     }
 }
 
-NTSTATUS NTAPI
-ConDrvWriteConsoleOutputString(IN PCONSOLE Console,
-                               IN PTEXTMODE_SCREEN_BUFFER Buffer,
-                               IN CODE_TYPE CodeType,
-                               IN PVOID StringBuffer,
-                               IN ULONG NumCodesToWrite,
-                               IN PCOORD WriteCoord,
-                               // OUT PCOORD EndCoord,
-                               OUT PULONG NumCodesWritten OPTIONAL)
+static NTSTATUS
+IntWriteConsoleOutputStringUnicode(
+    IN PCONSOLE Console,
+    IN PTEXTMODE_SCREEN_BUFFER Buffer,
+    IN PVOID StringBuffer,
+    IN ULONG NumCodesToWrite,
+    IN PCOORD WriteCoord,
+    OUT PULONG NumCodesWritten OPTIONAL)
 {
     NTSTATUS Status = STATUS_SUCCESS;
-    PVOID WriteBuffer = NULL;
-    PWCHAR tmpString = NULL;
-    ULONG X, Y, Length; // , Written = 0;
-    ULONG CodeSize;
+    LPWSTR WriteBuffer = StringBuffer;
+    ULONG i, X, Y, Length;
     PCHAR_INFO Ptr;
+    BOOL bCJK = Console->IsCJK;
 
-    if (Console == NULL || Buffer == NULL || WriteCoord == NULL /* || EndCoord == NULL */)
-    {
-        return STATUS_INVALID_PARAMETER;
-    }
-
-    /* Validity checks */
-    ASSERT(Console == Buffer->Header.Console);
-    ASSERT((StringBuffer != NULL) || (StringBuffer == NULL && NumCodesToWrite == 0));
-
-    //
-    // FIXME: Make overflow checks on WriteCoord !!!!!!
-    //
+    if (!StringBuffer)
+        goto Cleanup;
 
-    if (NumCodesWritten) *NumCodesWritten = 0;
+    X = WriteCoord->X;
+    Y = (WriteCoord->Y + Buffer->VirtualY) % Buffer->ScreenBufferSize.Y;
+    Length = NumCodesToWrite;
 
-    switch (CodeType)
+    for (i = 0; i < Length; ++i)
     {
-        case CODE_ASCII:
-            CodeSize = RTL_FIELD_SIZE(CODE_ELEMENT, AsciiChar);
-            break;
-
-        case CODE_UNICODE:
-            CodeSize = RTL_FIELD_SIZE(CODE_ELEMENT, UnicodeChar);
-            break;
-
-        case CODE_ATTRIBUTE:
-            CodeSize = RTL_FIELD_SIZE(CODE_ELEMENT, Attribute);
-            break;
+        Ptr = ConioCoordToPointer(Buffer, X, Y);
 
```

Comment:
Around there you should check, that if you are going to display a 2-byte width character but you are on a position `X` equal to `Buffer->ScreenBufferSize.X - 1`, you should go directly to a newline: indeed I'm not so sure it is possible to have a full-width character spanning a line-jump (you should compare with what happens on Windows).

Reply:
I will confirm.

Reply:
Win10 Japanese:
![jpn-echo-2](https://user-images.githubusercontent.com/2107452/72213527-18de1300-3534-11ea-958d-5fdaa3e53032.png)

Reply:
ReactOS Japanese with ReactOS JPN Package 0.7 from Rapps:
![ros](https://user-images.githubusercontent.com/2107452/72213598-e08b0480-3534-11ea-9dd9-5a1a0a100c64.png)